src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -589,17 +590,19 @@ private TypeGenerationSpec ParseTypeGenerationSpec(in TypeToGenerate typeToGener
                 }
 
                 var typeRef = new TypeRef(type);
-                string typeInfoPropertyName = typeToGenerate.TypeInfoPropertyName ?? GetTypeInfoPropertyName(type);
+                string typeInfoPropertyName = typeToGenerate.TypeInfoPropertyName ??
+                    type.ToIdentifierCompatibleSubstring(useUniqueName: false /* We leave identifier name conflicts to users, as codified below. */);
 
                 if (classType is ClassType.TypeUnsupportedBySourceGen)
                 {
                     ReportDiagnostic(DiagnosticDescriptors.TypeNotSupported, typeToGenerate.AttributeLocation ?? typeToGenerate.Location, type.ToDisplayString());
                 }
 
+                // Workaround for https://github.com/dotnet/roslyn/issues/54185 by keeping track of the file names we've used.
```

Comment:
The issue as described in https://github.com/dotnet/roslyn/issues/54185 hasn't been fixed in Roslyn 3.11 (and won't be fixed). We should keep this comment (and possibly with a TODO explaining that we should remove it once Roslyn 3.11 support is removed).